### PR TITLE
Add boundary constraints to the constraint graph

### PR DESCRIPTION
**Pull request by grjte**
This PR addresses part of #106, by moving boundary constraints into the `AlgebraicGraph` alongside the integrity constraints, and updating the codegen and tests accordingly.

It includes the following changes:

- [x] add the boundary statements to the graph
- [x] refactor to graph handling to simplify the combined handling of integrity + boundary constraints
- [x] update winter codegen to work with the new boundary constraints
- [x] update the tests to reflect the changes (primarily constraint re-ordering) and skip the tests which are failing due to the pending changes listed below

There are several things that yet to be completed, but I'd like to tackle these in a subsequent PR since this is already quite large:
 
- restore functionality of public inputs, by adding them to the graph, then restore the related tests
- restore boundary constraint validation in the IR, then restore the related tests
- add support for boundary variables
- verify that all cases are captured and add any additional validation/testing as required

### CODE REVIEW

File: air-script/tests/variables/variables.air
Comment (grjte) [inline air-script/tests/variables/variables.air:17]: I changed this to keep the public inputs usage in the public inputs test. Once we add variables for the boundary constraints, we can test them here as well

File: ir/src/constraints/mod.rs
Comment (Al-Kindi-0) [inline ir/src/constraints/mod.rs:74]: nit: to the first

File: ir/src/constraints/mod.rs
Comment (Al-Kindi-0) [inline ir/src/constraints/mod.rs:282]: nit: repeated "is"

**Review by Al-Kindi-0 — COMMENTED** [review]
Looks great to me, thanks @grjte !
Just some minor nits from my side.

File: ir/src/constraints/graph.rs
Comment (tohrnii) [inline ir/src/constraints/graph.rs:139]: nit: Should we define an alias `AUX_TRACE_SEGMENT`?

**Review by tohrnii — APPROVED** [review]
Thank you @grjte, looks good to me. I left a few minor nits inline.

**Review by Overcastan — APPROVED** [review]
Looks great to me, thank you @grjte!

File: codegen/winterfell/src/air/mod.rs
Comment (grjte) [reply codegen/winterfell/src/air/mod.rs:11]: Yeah, I don't like this name either, I just didn't have any better ideas at the time

File: codegen/winterfell/src/air/boundary_constraints.rs
Comment (grjte) [reply codegen/winterfell/src/air/boundary_constraints.rs:53]: I thought about that - it's how we did it before. It seemed a bit more complicated and maybe unnecessary, but I'll take another look. I don't really have a strong opinion either way

File: codegen/winterfell/src/air/boundary_constraints.rs
Comment (grjte) [reply codegen/winterfell/src/air/boundary_constraints.rs:57]: It's returning the root `NodeIndex` in the graph where the expression starts, which is why I called it `expr_root`. I don't have a strong opinion, but I think `expr` is less accurate

File: codegen/winterfell/src/air/mod.rs
Comment (grjte) [reply codegen/winterfell/src/air/mod.rs:119]: I'd prefer not to. I like seeing each step for each trace segment. That's how the codegen is set up in general, so it feels clearer and more explicit to me

File: ir/src/constraints/graph.rs
Comment (grjte) [reply ir/src/constraints/graph.rs:139]: Yeah, I think that makes sense. In the longer run we should make this more general, but that's very low priority, so we probably won't get there for a while

File: codegen/winterfell/src/air/boundary_constraints.rs
Comment (tohrnii) [reply codegen/winterfell/src/air/boundary_constraints.rs:57]: Makes sense. I was thinking that in this context it's clear that we are talking about the root and `expr` is more concise but I can see the argument to be explicit to make it clear.

File: codegen/winterfell/src/air/mod.rs
Comment (grjte) [reply codegen/winterfell/src/air/mod.rs:11]: I moved `ElemType`, but otherwise I haven't changed this. Basically, the problem with `codegen_trait` is that we actually have different codegen traits in a couple places (for constants and for periodic columns).  The problem with naming it something codegen-related is that everything in this crate is codegen-related, so I'm not sure that's useful. The idea of this name is that this is where the codegen is defined for the core types of the `ir`, but that's not even fully true. What is true is that it's where there's codegen that's related to the graph/constraints, so maybe I could change it to something about those?

Other thoughts/suggestions would be great

**Comment by grjte** [comment]
Thanks for your reviews @Al-Kindi-0 @Overcastan and @tohrnii! I made the suggested changes, except for one point of discussion raised by @tohrnii (which it would be great to get your thoughts on).

Since this is a very significant change, let's wait for the review from @bobbinth before we merge.

File: codegen/winterfell/src/air/boundary_constraints.rs
Comment (bobbinth) [inline codegen/winterfell/src/air/boundary_constraints.rs:66]: Thought for the future: it seems like determining if the `last_row` variable was declared or not adds quite a bit of complexity. Should we just always use `self.last_step()` instead? If it has `inline(always)` attribute, should theoretically give the same result.

File: ir/src/constraints/mod.rs
Comment (bobbinth) [inline ir/src/constraints/mod.rs:93]: As far as I can tell, this is a helper map which is used only during the construction of the graph and is not really needed after the graph is constructed. If so, it may be good to find a better way to handle it in the future so that once constructed `Constraints` does not need to retain any extra info.

File: ir/src/constraints/graph.rs
Comment (bobbinth) [inline ir/src/constraints/graph.rs:65]: Seems like this is used only in Winterfell codegen, and I'm not sure if this will be relevant in other codgens (or if it will be relevant in the same way). Maybe the logic should go into Winterfell codegen rather than being here?

**Review by bobbinth — APPROVED** [review]
Looks great! Thank you! I reviewed everything except for tests.

I think the structure is much cleaner than before. My comments are mostly for the future except fro the `split_boundary_constraint` comment. But even that one is non-blocking.

File: codegen/winterfell/src/air/mod.rs
Comment (tohrnii) [reply codegen/winterfell/src/air/mod.rs:11]: > What is true is that it's where there's codegen that's related to the graph/constraints, so maybe I could change it to something about those?

Makes sense. So, maybe just `graph` (seems to be inline with naming of other files)? I don't have a strong opinion though.

File: codegen/winterfell/src/air/mod.rs
Comment (grjte) [reply codegen/winterfell/src/air/mod.rs:32]: I was wondering about something along these lines as well, and I like this better, thanks!

File: codegen/winterfell/src/air/ir_types.rs
Comment (grjte) [reply codegen/winterfell/src/air/ir_types.rs:111]: Agreed. I was trying to minimize changes in this PR. I've opened #130 to track this.

Are you thinking this unrolling should be done in the IR or in the winterfell codegen? (It seems to make sense as a future optimization that could be done in the IR)

File: codegen/winterfell/src/air/boundary_constraints.rs
Comment (grjte) [reply codegen/winterfell/src/air/boundary_constraints.rs:66]: I'm happy to update this now. I actually originally declared `last_row` in all cases specifically to avoid this complexity, but this solution seems like the best option

File: ir/src/constraints/mod.rs
Comment (grjte) [reply ir/src/constraints/mod.rs:93]: Yes, this is true. I think there are still a few refactors that could be done, but I wanted to keep this PR as minimal as possible. Once the remaining issues related to this work are addressed this is worth revisiting (or possibly while addressing those)

File: codegen/winterfell/src/air/mod.rs
Comment (grjte) [reply codegen/winterfell/src/air/mod.rs:11]: Yeah, I think there might not be a great answer, but `graph` does seem better.

File: ir/src/constraints/graph.rs
Comment (grjte) [reply ir/src/constraints/graph.rs:65]: Yeah, I think you're right that it's better there. I've moved it.

**Pull request merged**
